packages/ceramic-core/src/document.ts

```diff
@@ -234,41 +242,20 @@ class Document extends EventEmitter {
    *
    * @param record - Record data
    * @param opts - Document initialization options (request anchor, wait, etc.)
-   * @param validate - Validate document against schema on apply
    */
-  async applyRecord (record: any, opts: DocOpts = {}, validate = true): Promise<void> {
+  async applyRecord (record: any, opts: DocOpts = {}): Promise<void> {
     const cid = await this.dispatcher.storeRecord(record)
 
-    const retrievedRec = await this.dispatcher.retrieveRecord(cid)
-    const state = await this._doctypeHandler.applyRecord(retrievedRec, cid, this._context, this.state)
-
-    if (DoctypeUtils.isAnchorRecord(retrievedRec)) {
-      state.anchorStatus = AnchorStatus.ANCHORED
-    }
-
-    let payload
-    if (retrievedRec.payload && retrievedRec.signatures) {
-      payload = (await this._context.ipfs.dag.get(retrievedRec.link)).value
-    } else {
-      payload = retrievedRec
-    }
-
-    if (payload.header) {
-      // override properties
-      Object.assign(state.metadata, payload.header)
-    }
-
-    if (validate) {
-      const schema = await Document.loadSchemaById(this._context.api, state.metadata.schema)
-      if (schema) {
-        Utils.validate(state.next.content, schema)
-      }
-    }
-
-    this._doctype.state = state
+    const blockPromise = new Promise((resolve, reject) => {
```

Comment:
Any details as to why there is this callback logic here please? Any reason the following wouldn't work?
```ts
await this._handleTip(cid)
await this._updateStateIfPinned()
await this._applyOpts(opts)
```

Reply:
I need to catch the error - ```await this._handleTip(cid)``` just logs it.

Reply:
If the error need to be caught by the caller then I think `_handleTip` should throw it, rather than having to use this kind of workaround that makes the logic harder to follow.

Reply:
the error needs to be caught just sometimes - on client interaction. This method is also binded as THE method that will process messages from pubsub as well. That's why we don't need to process all the errors from it.

```
this.on('update', this._handleTip.bind(this))
```

Reply:
Maybe the handler bound on the `update` event could be another method wrapping the `_handleTip` call in a `try {} catch {}` statement then so the error handling logic is always up to the consumer?

Reply:
Yep, that's one of the solutions. I didn't want to refactor that in this PR.

Reply:
@PaulLeCam refactored this as well 👍

Reply:
Nice!